File: 1_intro/3_smooth_sailing/10_commonCharacterCount/commonCharacterCount.js
function commonCharacterCount(s1, s2) {
  var alphabet1 = {};
  var alphabet2 = {};
  var output = 0;
  s1.split('').forEach(function(v) {
      if (v in alphabet1) {
          alphabet1[v]++;
      } else {
          alphabet1[v] = 1;
      }
  });
  s2.split('').forEach(function(v) {
      if (v in alphabet2) {
           alphabet2[v]++;
      } else {
          alphabet2[v] = 1;
      }
  });
  for (v in alphabet1) {
      if (v in alphabet2) {
          if (alphabet1[v] < alphabet2[v]) {
              output += alphabet1[v];
          } else {
              output += alphabet2[v];
          }
      }
  }
  return output;
}

File: 1_intro/3_smooth_sailing/10_commonCharacterCount/commonCharacterCount.lua
function commonCharacterCount(s1, s2)
  local alphabet1 = {}
  local alphabet2 = {}
  local output = 0
  stringCount(s1, alphabet1)
  stringCount(s2, alphabet2)
  for char, count in alphabet1 do
    if alphabet2[char] ~= nil then
      if alphabet1[char] < alphabet2[char] then
        output = output + alphabet1[char]
      else
        output = output + alphabet2[char]
      end
    end
  end
  return output
end

function stringCount(string, alphabet)
  for character in string:gmatch"." do
    if alphabet[character] ~= nil then
      alphabet[character] = alphabet[character] + 1
    else
      alphabet[character] = 1
    end
  end
end